File: packages/vintent/src/modules/csv/profiler.ts
export type FieldType = "nominal" | "ordinal" | "quantitative" | "temporal";

export interface DatasetProfile {
    fields: Record<
        string,
        {
            type: FieldType;
            cardinality: number;
        }
    >;
    rowCount: number;
}

/**
 * Given raw CSV text, compute a lightweight dataset profile.
 * The profile is used for validation and shell selection only.
 */
export function profileCsv(csvText: string): DatasetProfile {
    const lines = csvText
        .split(/\r?\n/)
        .map((line) => line.trim())
        .filter((line) => line.length > 0);

    if (lines.length === 0) {
        return { fields: {}, rowCount: 0 };
    }

    const headers = splitCsvLine(lines[0]);
    const rows = lines.slice(1).map(splitCsvLine);

    const fieldValues: Record<string, Set<string>> = {};
    const fieldTypes: Record<string, Set<"number" | "string" | "date">> = {};

    for (const header of headers) {
        fieldValues[header] = new Set();
        fieldTypes[header] = new Set();
    }

    for (const row of rows) {
        headers.forEach((header, index) => {
            const raw = row[index];
            if (raw == null || raw === "") {
                return;
            }

            fieldValues[header].add(raw);

            if (isNumeric(raw)) {
                fieldTypes[header].add("number");
            } else if (isDateLike(raw)) {
                fieldTypes[header].add("date");
            } else {
                fieldTypes[header].add("string");
            }
        });
    }

    const fields: DatasetProfile["fields"] = {};

    for (const header of headers) {
        const observedTypes = fieldTypes[header];
        const cardinality = fieldValues[header].size;

        let type: FieldType;

        if (observedTypes.size === 1 && observedTypes.has("number")) {
            type = "quantitative";
        } else if (observedTypes.size === 1 && observedTypes.has("date")) {
            type = "temporal";
        } else {
            type = "nominal";
        }

        fields[header] = {
            type,
            cardinality,
        };
    }

    return {
        fields,
        rowCount: rows.length,
    };
}

/**
 * Split a single CSV line into values.
 * Handles quoted values and commas inside quotes.
 */
function splitCsvLine(line: string): string[] {
    const result: string[] = [];
    let current = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const char = line[i];

        if (char === '"') {
            inQuotes = !inQuotes;
            continue;
        }

        if (char === "," && !inQuotes) {
            result.push(current);
            current = "";
        } else {
            current += char;
        }
    }

    result.push(current);
    return result.map((value) => value.trim());
}

/**
 * Check if a value is numeric.
 */
function isNumeric(value: string): boolean {
    return value !== "" && !isNaN(Number(value));
}

/**
 * Check if a value looks like a date.
 */
function isDateLike(value: string): boolean {
    const timestamp = Date.parse(value);
    return !isNaN(timestamp);
}

File: packages/vintent/vintent/vintent/core/modules/csv/profiler.py
import csv
import io
from datetime import datetime
from typing import Dict, Literal, Set, TypedDict
from ..schemas import FieldType, FieldInfo, DatasetProfile


def profile_csv(csv_text: str) -> DatasetProfile:
    """
    Given raw CSV text, compute a lightweight dataset profile.
    The profile is used for validation and shell selection only.
    """
    # Split lines and remove empty ones
    lines = [line.strip() for line in csv_text.strip().splitlines() if line.strip()]

    if not lines:
        return {"fields": {}, "rowCount": 0}

    # Parse CSV properly using Python's csv module
    csv_reader = csv.reader(io.StringIO(csv_text.strip()))
    rows = list(csv_reader)

    if not rows:
        return {"fields": {}, "rowCount": 0}

    headers = rows[0]
    data_rows = rows[1:]

    field_values: Dict[str, Set[str]] = {header: set() for header in headers}
    field_types: Dict[str, Set[str]] = {header: set() for header in headers}

    for row in data_rows:
        # Ensure row has same number of columns as headers (pad with empty strings if needed)
        row = list(row) + [""] * (len(headers) - len(row))

        for header, raw in zip(headers, row):
            if raw is None or raw == "":
                continue

            field_values[header].add(raw)

            if is_numeric(raw):
                field_types[header].add("number")
            elif is_date_like(raw):
                field_types[header].add("date")
            else:
                field_types[header].add("string")

    fields: Dict[str, FieldInfo] = {}

    for header in headers:
        observed_types = field_types[header]
        cardinality = len(field_values[header])

        if len(observed_types) == 1 and "number" in observed_types:
            field_type: FieldType = "quantitative"
        elif len(observed_types) == 1 and "date" in observed_types:
            field_type = "temporal"
        else:
            # Note: Python version only distinguishes nominal vs others
            # You might want to add ordinal detection logic here
            field_type = "nominal"

        fields[header] = {"type": field_type, "cardinality": cardinality}

    return {"fields": fields, "rowCount": len(data_rows)}


def is_numeric(value: str) -> bool:
    """Check if a value is numeric."""
    try:
        float(value)
        return True
    except (ValueError, TypeError):
        return False


def is_date_like(value: str) -> bool:
    """Check if a value looks like a date."""
    # Try multiple common date formats
    date_formats = [
        "%Y-%m-%d",
        "%Y/%m/%d",
        "%d-%m-%Y",
        "%d/%m/%Y",
        "%m-%d-%Y",
        "%m/%d/%Y",
        "%Y-%m-%d %H:%M:%S",
        "%Y/%m/%d %H:%M:%S",
        "%d-%m-%Y %H:%M:%S",
        "%d/%m/%Y %H:%M:%S",
    ]

    for fmt in date_formats:
        try:
            datetime.strptime(value, fmt)
            return True
        except ValueError:
            continue

    # Also try the generic date parsing
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
        return True
    except ValueError:
        pass

    return False
